### PR TITLE
Add to_dataframe to InferenceData

### PR DESCRIPTION
**Pull request by ahartikainen**
## Description

Add `.to_dataframe` function to InferenceData. The result is a dataframe in a "wide" format.

This PR also adds `__getitem__` magic for the InferenceData.

## Checklist

- [x] Follows [official](https://github.com/arviz-devs/arviz/blob/master/CONTRIBUTING.md#pull-request-checklist) PR format
- [x] Includes a sample plot to visually illustrate the changes (only for plot-related functions)
- [ ] New features are properly documented (with an example if appropriate)?
- [x] Includes new or updated tests to cover the new feature
- [x] Code style  correct (follows pylint and black guidelines)
- [x] Changes are listed in [changelog](https://github.com/arviz-devs/arviz/blob/master/CHANGELOG.md#v0xx-unreleased)

image

image

image

image

image

### CODE REVIEW

File: arviz/data/inference_data.py [hunk 1]
```diff
@@ -190,6 +190,12 @@ def __iter__(self):
         for group in self._groups_all:
             yield group
 
+    def __getitem__(self, key):
+        """Get item by key."""
+        if key not in self._groups_all:
+            raise KeyError(key)
+        return getattr(self, key)
```
Comment (OriolAbril) [inline arviz/data/inference_data.py:197]: :heart:

File: arviz/data/inference_data.py [hunk 2]
```diff
@@ -343,13 +363,156 @@ def to_json(self, filename, **kwargs):
...
+        Returns
+        -------
+        pandas.DataFrame
+            A pandas DataFrame containing all selected groups of InferenceData object.
+        """
+        # pylint: disable=too-many-nested-blocks
+        if not include_coords and not include_index:
+            raise TypeError("Both include_coords and include_index can not be False.")
+        if index_origin not in [0, 1]:
+            raise TypeError("index_origin must be 0 or 1, saw {}".format(index_origin))
```
Comment (OriolAbril) [inline arviz/data/inference_data.py:417]: should this take the default from rcParams?

File: arviz/data/inference_data.py [hunk 3]
```diff
@@ -343,13 +363,156 @@ def to_json(self, filename, **kwargs):
...
+        if groups is None:
```
Comment (OriolAbril) [inline arviz/data/inference_data.py:418]: I think calling `_group_names` and filtering for `data` afterwards will be enough.

If `groups` passed to `_group_names` is `None` it automatically uses `_groups_all`, if it's a string or list of strings it expands the metagroups present and returns the matching groups.

File: arviz/data/inference_data.py [hunk 4]
```diff
@@ -343,13 +363,156 @@ def to_json(self, filename, **kwargs):
...
+                filter(lambda x: "data" not in x, self._group_names(groups, filter_groups))
+            )
+
+        if any(
+            group in ["observed_data", "constant_data", "predictions_constant_data"]
+            for group in groups
+        ):
+            raise ValueError("Cannot transform data groups to dataframe")
+
+        if not all(group in self._groups_all for group in groups):
```
Comment (OriolAbril) [inline arviz/data/inference_data.py:450]: `_group_names` should also check for this

**Review by OriolAbril — COMMENTED** [review]
LGTM, thanks!

File: arviz/data/inference_data.py (hunk 2 above)
Comment (ahartikainen) [reply arviz/data/inference_data.py:417]: It should

File: arviz/data/inference_data.py [hunk 5]
```diff
@@ -343,13 +363,132 @@ def to_json(self, filename, **kwargs):
...
+            index_origin = rcParams["data.index_origin"]
+        if index_origin not in [0, 1]:
+            raise TypeError("index_origin must be 0 or 1, saw {}".format(index_origin))
+
+        group_names = list(
+            filter(lambda x: "data" not in x, self._group_names(groups, filter_groups))
+        )
+
+        if not group_names:
+            raise TypeError("No valid groups found: {}".format(groups))
```
Comment (ahartikainen) [inline arviz/data/inference_data.py:431]: Is this logic ok?

posterior + observed_data --> df from posterior

observed_data -> TypeError

Comment (OriolAbril) [reply arviz/data/inference_data.py:431]: I think it is ok, given that it fits our goals and we document it. My only question not really knowing yet how panel works is if ppc checks will be possible with it if observed_data is not in the dataframe

Comment (ahartikainen) [reply arviz/data/inference_data.py:431]: Panel works with array too. My gist was just short cutting some code with data frame.

Comment (OriolAbril) [reply arviz/data/inference_data.py:431]: Oh, cool!

**Pull request merged**